app/models/prisoner_availability_validation.rb

```diff
@@ -21,46 +27,43 @@ def unknown_result?
   attr_reader :api_error
 
   def slots_availability
-    valid_requested_dates.each do |requested_date|
-      error_message = error_message_for_slot(requested_date)
-      errors[requested_date.to_s] << error_message if error_message
+    valid_requested_slots.each do |requested_slot|
+      error_messages_for_slot(requested_slot).each do |message|
+        errors[requested_slot.to_s] << message
+      end
     end
   end
 
-  def error_message_for_slot(date)
-    return if unknown_result?
+  def error_messages_for_slot(slot)
+    return [] if unknown_result? || !valid_slot?(slot)
 
-    PRISONER_NOT_AVAILABLE unless offender_availability.include?(date)
+    offender_availability.error_messages_for_slot(slot)
   end
 
   def offender_availability
     return nil unless offender.valid?
-    # Don't want to show prisoner unavailable errors for invalid dates
-    return requested_dates unless valid_requested_dates.any?
 
     @offender_availability ||= load_offender_availability
   end
 
   def load_offender_availability
     return nil if @api_error
 
-    Nomis::Api.instance.offender_visiting_availability(
+    Nomis::Api.instance.offender_visiting_detailed_availability(
       offender_id: offender.id,
-      start_date:  valid_requested_dates.min,
-      end_date:    valid_requested_dates.max
-    ).dates
+      slots: valid_requested_slots
+    )
   rescue Nomis::APIError => e
     @api_error = true
     Rails.logger.warn "Error calling the NOMIS API: #{e.inspect}"
     nil
   end
 
-  def valid_requested_dates
-    @valid_dates ||=
-      begin
-        min = Date.tomorrow
-        max = 60.days.from_now.to_date
-        requested_dates.select { |date| date.between?(min, max) }
-      end
+  def valid_requested_slots
+    @valid_slots ||= requested_slots.select { |slot| valid_slot?(slot) }
+  end
+
+  def valid_slot?(slot)
+    slot.to_date.between?(Date.tomorrow, 60.days.from_now.to_date)
```

Comment:
Agree in principle, although I'm not sure if this is going to make things clearer.

The tomorrow constraint is because we don't want to consider dates of today, although the api accepts today, and the 60 days is an API requirement, we could put them as constants but then we would be adding indirection and we would lose clarity on what a valid slot is easily as the constants would be in different classes. I think in this case since these numbers are only used in this method it is actually not an issue. What do you think?